akka-docs/rst/java/persistence.rst

```diff
@@ -775,6 +775,29 @@ A snapshot of state data can be persisted by calling the ``saveStateSnapshot()``
 On recovery state data is initialized according to the latest available snapshot, then the remaining domain events are replayed, triggering the
 ``applyEvent`` method.
 
+
+Periodical snapshot
+-------------------
+
+You can enable periodical ``saveStateSnapshot()`` calls in ``PersistentFSM`` if you turn the following flag on in ``reference.conf``::
+
+    akka.persistence.fsm.enable-snapshot-after = on
+
+and also configure the ``snapshot-after`` value. For example if you specify::
+
+    akka.persistence.fsm.snapshot-after = 1000
+
+this means ``saveStateSnapshot()`` is called when the sequence number is multiple of 1000.
+
+.. note::
+
+ When you are calling ``applying`` method in the ``PersistentFSM`` DSL by passing multiple domain events,
+ then ``saveStateSnapshot()`` might not be called exactly at sequence numbers being multiple of the
+ ``snapshot-after`` configuration value.
+ This is because when ``applying`` takes multiple domain events, they are handled in a sort of "batch"
+ mode and  ``saveStateSnapshot()`` is called only at the end of batch.
```

Comment:
>Isn't there one event for the state transition and one (or more) domain events? 

Yes, that is true. 

>Could that result in that you accidentally get into a series of sequence numbers that never match the modulo condition?
E.g. always odd numbers will never trigger snapshots after 1000.
Could we handle that somehow?

Sorry I am afraid I failed to understand this... `saveStateSnapshot()` might not be called exactly at `lastSequenceNr` being multiple of 1000. However, it is still called **after** multiple of 1000 persistent events (maybe at 1003, 1004, etc) . My [test case](https://github.com/richard-imaoka/akka/blob/e49416b70e649b94701e8d48a2bdd75259ec27f3/akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMSpec.scala#L374) shows that (SeqNo = 8, not multiple of 3). 

Below is what I investigated - Let me know if I got your question wrong.:

* `PersistentFSM`'s [applyState](https://github.com/richard-imaoka/akka/blob/e49416b70e649b94701e8d48a2bdd75259ec27f3/akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala#L115L117), does below

```scala
    if (nextState.notifies || nextState.timeout.nonEmpty) {
      eventsToPersist = eventsToPersist :+ StateChangeEvent(nextState.stateName.identifier, nextState.timeout)
    }
```

* and [`persistAll`](https://github.com/richard-imaoka/akka/blob/e49416b70e649b94701e8d48a2bdd75259ec27f3/akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala#L146) is called for `eventsToPersist`

```scala
      persistAll[Any](eventsToPersist) { ... }
```

* That means every single event in `eventsToPersist`, regardless of [domain event](https://github.com/richard-imaoka/akka/blob/e49416b70e649b94701e8d48a2bdd75259ec27f3/akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala#L149) or [state transition](https://github.com/richard-imaoka/akka/blob/e49416b70e649b94701e8d48a2bdd75259ec27f3/akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala#L152), will trigger persistence handler and the modulo condition is checked for every single event